src/common/s3.connection.hh

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <condition_variable>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+
+namespace minio::s3 {
+class Client;
+class Part;
+} // namespace minio::s3
+
+namespace minio::creds {
+class StaticProvider;
+} // namespace minio::creds
+
+namespace acquire::sink::zarr::common {
+struct S3Connection final
+{
+    S3Connection(const S3Connection&) = delete;
+    explicit S3Connection(const std::string& endpoint,
+                          const std::string& access_key_id,
+                          const std::string& secret_access_key);
+
+    ~S3Connection() noexcept;
+
+    // bucket operations
+    bool bucket_exists(const std::string& bucket_name);
+    [[nodiscard]] bool make_bucket(const std::string& bucket_name);
```

Comment:
I'd made the assumption that, because Zarr datasets could so large with so many files, each dataset would basically correspond to a bucket -- but now you mention it, that probably be the case. We would want to be able to serve users who can write to a bucket but not necessarily create one.

Reply:
I moved them to an anonymous namespace for tests, realized I didn't need them, forgot to remove them (I was in a hurry before standup).